Pare down README so it applies to both the service and base source

cmcantalupo opened the pull request:
- Fixes #1666 from github issues.

bgeltz commented on RUNTIME.md:
```diff
@@ -0,0 +1,536 @@
+
+
+
+
+
```
```suggestion
```

bgeltz commented on README:
```diff
@@ -598,16 +138,19 @@ GEOPM with the `--enable-beta` configure flag.  The features currently
-2020 September 25
+2021 September 24
 
 Christopher Cantalupo <christopher.m.cantalupo@intel.com> <br>
```
```Brad Geltz <brad.geltz@intel.com> <br>```

bgeltz commented on README:
```diff
@@ -32,553 +37,88 @@ With GEOPM you can:
-    ./configure --prefix=$HOME/build/geopm
-    make
-    make install
+GEOPM is an open development project and we use github to plan, review
```
```suggestion
GEOPM is an open development project and we use GitHub to plan, review
```

bgeltz commented on RUNTIME.md:
```diff
@@ -0,0 +1,536 @@
+resources for multiple MPI jobs and multiple users by the system
+resource manager.
+
+The GEOPM package provides two libraries: libgeopm for use with MPI
```
What about libgeopmd?  There's no discussion of the service here.

cmcantalupo replied:
I've been trying to refer to the "GEOPM package" and the "GEOPM Service package" perhaps we should call the base build package the "GEOPM HPC Runtime package" throughout?

bgeltz replied:
I like being explicit about the "GEOPM HPC package" or "...runtime package"

bgeltz commented on RUNTIME.md:
```diff
@@ -0,0 +1,536 @@
+package.  For example, to install the geopmpy package into your home
+directory, run the following command:
+
+    python -m pip install --user geopmpy
```
Also geopmdpy

bgeltz commented on RUNTIME.md:
```diff
@@ -0,0 +1,536 @@
+
+INSTALL
+-------
+The OpenHPC project provides the most robust way to install GEOPM.
```
The version in OpenHPC does not have any of the service bits.  Not sure this statement is true anymore.

bgeltz commented on RUNTIME.md:
```diff
@@ -0,0 +1,536 @@
+install the binaries required by the GEOPM runtime.  Without these binary files
+*only* the following modules will load successfully:
+
+    geopmpy.io
+    geopmpy.launcher
+    geopmpy.plotter
+    geopmpy.update_report
```
We should check that this is still correct.

bgeltz commented on RUNTIME.md:
```diff
@@ -0,0 +1,536 @@
+    geopmpy.update_report
+
+All other modules rely on the output of the GEOPM build (namely
+libgeopmpolicy.so) being present and available in LD_LIBRARY_PATH or
```
No discussion of libgeopmd.so dependency.

bgeltz commented on RUNTIME.md:
```diff
@@ -0,0 +1,536 @@
+rubygems and ruby-devel package on RHEL and SLES, or ruby and ruby-dev on
+Ubuntu.
+
+RHEL:
+
+    yum install openmpi-devel elfutils-libelf-devel ruby-devel rubygems
+
+SUSE:
+
+    zypper install openmpi-devel elfutils-libelf-devel ruby-devel rubygems
+
+UBUNTU (as of 18.04.3 LTS):
+
+    apt install libtool automake libopenmpi-dev build-essential gfortran \
+        libelf-dev ruby ruby-dev python libsqlite3-dev
```
Missing reqs for service (e.g. python-gobject).

bgeltz commented on RUNTIME.md:
```diff
@@ -0,0 +1,536 @@
+options.
+
+
+BUILD INSTRUCTIONS
+------------------
```
This whole section needs to be reworked to discuss the build script and/or the environment files in integration/config that setup CC, CXX, etc. for a given system.

bgeltz commented on RUNTIME.md:
```diff
@@ -0,0 +1,536 @@
+    make
+    make install
+
+An RPM can be created on a RHEL or SUSE system with the
+
+    make rpm
```
We have a whole other guide on how to build RPMs now.  We should just point to that.

bgeltz commented on RUNTIME.md:
```diff
@@ -0,0 +1,536 @@
+
+RUN REQUIREMENTS
+----------------
+We are targeting SLES12 and RHEL7 distributions for functional runtime
```
Old OS versions discussed.  This isn't accurate today.

bgeltz commented on RUNTIME.md:
```diff
@@ -0,0 +1,536 @@
+
+For additional information, please contact the GEOPM team.
+
+### USER ENVIRONMENT
```
This should be reworked to discuss the ```integration/config/run_env.sh``` script.

bgeltz commented on RUNTIME.md:
```diff
@@ -0,0 +1,536 @@
+
+TESTING
+-------
+From within the source code directory, unit tests can be executed with
```
Which source code directory?  The base or service?

We should be explicit about their being two different spots that you need to do this process in.  Examples would be nice too.

bgeltz replied:
This should be under a sub heading for "Unit Tests"

bgeltz commented on RUNTIME.md:
```diff
@@ -0,0 +1,536 @@
+The unit tests can be executed on any development system, including
+VMs and containers, that meets the BUILD REQUIREMENTS section above.
+
+The integration tests are located in the "integration/test" directory.
```
This should be under a sub heading for "Integration Tests"

bgeltz commented on RUNTIME.md:
```diff
@@ -0,0 +1,536 @@
+python package to validate the runtime.  Please report failures of
+these tests as issues.
+
+Note that if the `GEOPM_SKIP_COMPILER_CHECK` environment variable is
+set when `${GEOPM_SOURCE}/build.sh` is called, then the check to
+verify that the Intel compilers are used will be bypassed.  This can
+be useful for performing builds with the GNU toolchain.
```
This should be moved to the build section.  Not sure why it's here.  Not relevant to integration tests.

bgeltz commented on RUNTIME.md:
```diff
@@ -0,0 +1,536 @@
+verify that the Intel compilers are used will be bypassed.  This can
+be useful for performing builds with the GNU toolchain.
+
+RESOURCE MANAGER INTEGRATION
```
Are there implications of the service and libgeopmd.so that we have not captured in this section?

e.g. do we need libgeopmd.so to make these calls to save_control, etc.?

bgeltz commented on service/docs/source/devel.rst:
```diff
@@ -3,7 +3,11 @@ Guide for GEOPM Developers
+source code in the GEOPM repository.  The GEOPM repository contains
+two independent build systems that are used to compile, test and
+install software components written in C++ and Python.  The `service`
+subdirectory of the geopm repository contains a build system for the
```
```suggestion
subdirectory of the GEOPM repository contains a build system for the
```

bgeltz commented on service/docs/source/devel.rst:
```diff
@@ -2,5 +2,123 @@
+
+The integration tests require a system with MPI and multiple compute nodes.
+
+Coverage Instructions
+---------------------
```
Check out what I started here:
https://github.com/geopm/geopm/pull/1893/files

bgeltz commented on build.sh:
```diff
@@ -72,19 +72,25 @@ if [ ! -f "${BUILD_ENV}" ]; then
+fi
+
+# Clear out old builds unless we are skipping install
+if [ -d "${GEOPM_INSTALL}" ] && [ -z "${GEOPM_SKIP_INSTALL+x}" ]; then
```
This seems risky.  I have absolutely debugged issues before where a stale build was being picked up because it was in the LD_LIBRARY_PATH o and/or PATH, and the current build was never installed.  I'd really prefer to just remove any old build that's laying around to avoid problems like this.

cmcantalupo replied:
It seems really strange to error because GEOPM_INSTALL is not defined but GEOPM_SKIP_INSTALL is defined. If the user has chosen to "skip install" why would we want call rm -rf on whatever happens to by living in the environment variable "GEOPM_INSTALL"?  Perhaps we should add a prompt asking the user before force removing an entire directory?  I don't like it when rm -rf applies to a variable that I'm not using.

bgeltz commented on build.sh:
```diff
@@ -72,19 +72,25 @@ if [ ! -f "${BUILD_ENV}" ]; then
 
-# Clear out old builds
-if [ -d "${GEOPM_INSTALL}" ]; then
+# Check that GEOPM_INSTALL is defined unless we are skipping install
+if [ -z "${GEOPM_INSTALL}" ] && [ -z "${GEOPM_SKIP_INSTALL+x}" ];
+    echo "Please define GEOPM_INSTALL prior to using this script" 1>&2
+    exit 1
+fi
```
If they're using the recommendations from the docs, then they're using ~/.geopmrc and GEOPM_INSTALL is set.  Even if we aren't installing, having GEOPM_INSTALL set is benign.  Why bother?

Additionally, I don't think this is even reachable if GEOPM_INSTALL is not set.  build_env.sh was sourced on line 73, and enforces that GEOPM_INSTALL is set:
https://github.com/geopm/geopm/blob/dev/integration/config/build_env.sh#L49

bgeltz commented on build.sh:
```diff
@@ -72,19 +72,25 @@ if [ ! -f "${BUILD_ENV}" ]; then
 
-# e.g. To do a debug build, add --enable-debug to this env var
-GEOPM_GLOBAL_CONFIG_OPTIONS="${GEOPM_GLOBAL_CONFIG_OPTIONS} --prefix=${GEOPM_INSTALL}"
+# Append prefix to global config options unless skipping install
+if [ -z "${GEOPM_SKIP_INSTALL+x}" ]; then
+    GEOPM_GLOBAL_CONFIG_OPTIONS="${GEOPM_GLOBAL_CONFIG_OPTIONS} --prefix=${GEOPM_INSTALL}"
+fi
```
Having the prefix set when we are not installing should not affect anything, right?  Why bother?

cmcantalupo replied:
It goes in the rpath of the compiled binaries.

cmcantalupo commented on service/docs/source/devel.rst:
```diff
@@ -3,59 +3,236 @@ Guide for GEOPM Developers
+autotools package.
+
+Running the configure scripts generate a number of output files,
+including the ``Makefile`` that is use for the rest of the build
```
use -> used

cmcantalupo commented on service/docs/source/devel.rst:
```diff
@@ -3,59 +3,236 @@ Guide for GEOPM Developers
+Both configure scripts
+^^^^^^^^^^^^^^^^^^^^^^
+
+* ``--enable-debug``
```
document --prefix

bgeltz commented on integration/config/build.sh:
```diff
@@ -74,19 +74,25 @@ if [ ! -f "${BUILD_ENV}" ]; then
-# Clear out old builds
-if [ -d "${GEOPM_INSTALL}" ]; then
+# Check that GEOPM_INSTALL is defined unless we are skipping install
+if [ -z "${GEOPM_INSTALL}" ] && [ -z "${GEOPM_SKIP_INSTALL+x}" ];
```
Continuing from: https://github.com/geopm/geopm/pull/1898#discussion_r719458258
It's impossible to get here if ```${GEOPM_INSTALL}``` is not defined.  This is enforced by ```set -e```  and the sourcing of ```build_env.sh``` above.  If the user has chosen to skip the install, we want to be absolutely certain that we remove anything that's in ```GEOPM_INSTALL``` so it is not picked up.  We have a number of different scripts that will add to a user's ```PATH``` and ```LD_LIBRARY_PATH```, so in order to avoid debugging problems with stale builds it seems to be the most safe to just remove an old build.

I'm fine to add a check around the ```rm -fr``` to be sure they want it removed.  But if they don't want it removed, then I want to ```exit 1``` here and stop.  The user should be forced to move the old build from ```$GEOPM_INSTALL``` to some other backup directory.   Having ```$GEOPM_INSTALL``` contain an old build that's not currently under test is a recipe for debugging headaches.

bgeltz commented on integration/config/build.sh:
```diff
@@ -74,19 +74,25 @@ if [ ! -f "${BUILD_ENV}" ]; then
 
-# e.g. To do a debug build, add --enable-debug to this env var
-GEOPM_GLOBAL_CONFIG_OPTIONS="${GEOPM_GLOBAL_CONFIG_OPTIONS} --prefix=${GEOPM_INSTALL}"
+# Append prefix to global config options unless skipping install
+if [ -z "${GEOPM_SKIP_INSTALL+x}" ]; then
+    GEOPM_GLOBAL_CONFIG_OPTIONS="${GEOPM_GLOBAL_CONFIG_OPTIONS} --prefix=${GEOPM_INSTALL}"
+fi
```
Continuing from: https://github.com/geopm/geopm/pull/1898#discussion_r719460267
Is that a problem?  If we enforce that anything that was in ```$GEOPM_INSTALL``` was removed above, is there another issue here?

bgeltz commented on service/docs/source/devel.rst:
```diff
@@ -3,59 +3,213 @@ Guide for GEOPM Developers
+    # Build the geopm-service package
+    cd service/
+    ./autogen.sh
+    ./configure
```
If systemd is default on still (I think it is), this will be broken on systems that do not have the systemd dependencies from the package manager.

Since most users won't have access to install these requirements on large systems (e.g. Theta) I think we should default to systemd disabled.

bgeltz replied:
Open issue on investigating this.

bgeltz replied:
Default off until the next release?

cmcantalupo replied:
I added a lot of documentation that should help if the user needs to disable access to the service in when building the service.

bgeltz commented on service/docs/source/devel.rst:
```diff
@@ -3,59 +3,213 @@ Guide for GEOPM Developers
-   ./configure --disable-mpi
-   make
-   make check
+Service configure script
```
The requirement to build with GNU is not captured.

bgeltz commented on service/docs/source/devel.rst:
```diff
@@ -3,59 +3,213 @@ Guide for GEOPM Developers
+* ``export MPIFC=``
+  Set the Fortran compiler wrapper with environment variable
+
+* ``export MPIF77=``
+  Set the Fortran 77 compiler wrapper with environment variable
```
Also ~MPIF90, MPIFORT~, F90, F77, FC...

bgeltz commented on integration/config/build.sh:
```diff
@@ -74,19 +74,25 @@ if [ ! -f "${BUILD_ENV}" ]; then
 
-# Clear out old builds
-if [ -d "${GEOPM_INSTALL}" ]; then
+# Check that GEOPM_INSTALL is defined unless we are skipping install
```
A small comment saying this is an additional check for GEOPM_INSTALL above and beyond what's happening in the build_env.sh would be nice.

bgeltz replied:
And some testing of all the different combinations of these env variables.

bgeltz commented on service/docs/source/devel.rst:
```diff
@@ -3,59 +3,213 @@ Guide for GEOPM Developers
-   ./configure --disable-mpi
-   make
-   make check
+Service configure script
```
The option to ```--disable-systemd``` should be high in this list since it is most likely to be used on older systems.

bgeltz replied:
The option to ```--disable-systemd``` should be high in this list since it is most likely to be used on older systems.

bgeltz commented on RUNTIME.md:
```diff
@@ -0,0 +1,337 @@
+[geopm(7)](https://geopm.github.io/man/geopm.7.html).
+
+The GEOPM runtime provides some build in algorithms, each is called an
+"Agent" that implements the geopm::Agent class interface.  A developer
```
Man page reference and links?

bgeltz commented on RUNTIME.md:
```diff
@@ -0,0 +1,337 @@
+applications to enable application feedback for informing the control
+decisions.  The indirect calls are facilitated by GEOPM's integration
+with MPI and OpenMP through their profiling decorators, and the direct
+calls are made through the geopm_prof_c(3) or geopm_fortran(3)
```
Man pages links?

cmcantalupo replied:
These are all linked from the geopm(7) web page which has an explicit hyperlink in the rst file.  Perhaps we can open an issue to deal with the multitude of references to man pages that do not have hyperlinks.

bgeltz commented on RUNTIME.md:
```diff
@@ -0,0 +1,337 @@
+interfaces.  Marking up a compute application with profiling
+information through these interfaces can enable better integration of
+the GEOPM runtime with the compute application and more precise
+control.
```
Tutorial plug?

bgeltz commented on RUNTIME.md:
```diff
@@ -0,0 +1,337 @@
+For additional information, please contact the GEOPM team.
+
+
+### User Environment
```
Does this info still belong here since it duplicates part of what's discussed in the devel.rst?

bgeltz commented on RUNTIME.md:
```diff
@@ -0,0 +1,337 @@
+Use a PYTHONPATH that points to the site-packages created by the geopm build.
+The version created is for whichever version of python 3 was used in the
+configure step.  If a different version of python is desired, override the
+default with the --with-python option in the configure script.
```
Configure option to devel.rst?

bgeltz commented on RUNTIME.md:
```diff
@@ -0,0 +1,337 @@
+installed using the RPMs distributed there.  The source code for the
+driver can be found here at the link below.
+
+[msr-safe repo](https://github.com/scalability-llnl/msr-safe)
```
Repo URL needs to be updated.

bgeltz commented on configure.ac:
```diff
@@ -218,6 +218,28 @@ fi
 AC_SUBST([enable_fortran])
 AM_CONDITIONAL([ENABLE_FORTRAN], [test "x$enable_fortran" = "x1"])
 
+AC_ARG_ENABLE([geopmd-local],
```
Comment on ordering of this option vs. with-geopmd below being vital to the logic functioning correctly here.

cmcantalupo committed:
Provide default option for finding geopmd

- If the user does an in place build of the service
  directory, it will be used unless --disable-geopmd-local
  is provided at configure time.
- If the user specifies --with-geopmd options, these
  will override the default as well.

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Remove "lib" string from install prefix configure options

- The configure parameters for nvml and levelzero define the
  install root.
- This directory contains the "lib" directory and the "include"
  directory.
- Using "lib" in the command line option is confusing since it is
  not the path to the "lib" directory, but rather the directory
  that contains it.
- This changes --with-liblevelzero to --with-levelzero and
  --with-libnvml to --with-nvml.
- If we need to specify the lib and include directories distinctly
  in the future we may have --with-levelzero-lib and
  --with-levelzero-include as configure options.

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Move the build.sh script into integration/config

- The script relies heavily on other bash scripts in that
  directory.
- Clean up use of GEOPM_SOURCE and GEOPM_INSTALL
- GEOPM_SOURCE is only required when build.sh is not
  run from the root of the repo.
- GEOPM_INSTALL is not required if GEOPM_SKIP_INSTALL is
  set.

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Move most of README content into sphinx web page

- Improve developer and build documentation.
- Add runtime documentation to sphinx web page
- Update content to reflect the service

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

bgeltz commented on service/docs/source/devel.rst:
```diff
@@ -21,7 +21,7 @@ module.
 -----------------------
 
-The simplest procedure for building all of the software in the GEOPM
+The basic procedure for building all of the software in the GEOPM
```
:)

bgeltz commented on service/docs/source/requires.rst:
```diff
@@ -40,77 +59,75 @@ the requirement must be removed from the file
+There are some time-of-use requirements for both the GEOPM Service and
+the GEOPM HPC Runtime.  The use of MSRs enables many important
+features of the GEOPM Service, although access to MSRs is strictly
+required only for the GEOPM HPC Runtime.  There may be some
```
```suggestion
features of the GEOPM Service, although access to MSRs is strictly
required only for the GEOPM HPC Runtime, gathering telemetry through
``geopmread``, setting hardware controls with ``geopmwrite``, or using
GEOPM's Python APIs to achieve similar functionality as ``geopmread`` and ``geopmwrite``.  There may be some
```
